src/Simulator/IncrementalRedstoneSimulator.cpp

```diff
@@ -503,152 +514,280 @@ void cIncrementalRedstoneSimulator::HandleRedstoneButton(int a_RelBlockX, int a_
+					}
+					SetBlockPowered(AdjustedOffset, RelBlock, MyPower);
+				}
+				case E_BLOCK_REDSTONE_TORCH_ON:
+				case E_BLOCK_REDSTONE_WIRE:
+				{
+					BorderingMechanismCount++;
+					if (!RepeaterPresent)
+					{
+						OffsetToPower = { -Offset.x, Offset.y, -Offset.z };
+					}
+				}
+				default: break;
 			}
 		}
-		else if ((i >= 8) && (i <= 11))  // See above, but this is for wire below us
+	}
+
+	if (BorderingMechanismCount == 0)
+	{
+		SetAllDirsAsPowered(a_RelBlockX, a_RelBlockY, a_RelBlockZ);
+		for (const auto & BlockFaceOffset : BlockFaceOffsets)
 		{
-			BLOCKTYPE Type = 0;
-			if (!m_Chunk->UnboundedRelGetBlockType(a_RelBlockX + gCrossCoords[i].x, a_RelBlockY, a_RelBlockZ + gCrossCoords[i].z, Type))
-			{
-				continue;
-			}
-			if (cBlockInfo::IsSolid(Type))
-			{
-				continue;
-			}
+			SetDirectionLinkedPowered(a_RelBlockX, a_RelBlockY, a_RelBlockZ, BlockFaceOffset, MyPower);
 		}
+	}
+	else if (BorderingMechanismCount == 1)
+	{
+		eBlockFace Face = BlockFaceOffsets[std::distance(VectorOffsets, std::find(VectorOffsets, VectorOffsets + ARRAYCOUNT(VectorOffsets), OffsetToPower))];
+		SetBlockPowered(RelBlock + OffsetToPower, RelBlock, MyPower);
+		SetDirectionLinkedPowered(a_RelBlockX, a_RelBlockY, a_RelBlockZ, Face, MyPower);
+	}
+}
+
+
+
+
+
+void cIncrementalRedstoneSimulator::FindAndPowerBorderingWires(std::vector<std::pair<Vector3i, cChunk *>> & a_PotentialWireList, const std::pair<Vector3i, cChunk *> & a_Entry)
+{
+	static const Vector3i LevelOffsets[] =  // Wires on same level
+	{
+		{ 1, 0, 0 },
+		{ -1, 0, 0 },
+		{ 0, 0, 1 },
+		{ 0, 0, -1 },
+	};
+	static const Vector3i HigherOffsets[] =  // Wires one higher, surrounding self
+	{
+		{ 1, 1, 0 },
+		{ -1, 1, 0 },
+		{ 0, 1, 1 },
+		{ 0, 1, -1 },
+	};
+	static const Vector3i LowerOffsets[] =  // Wires one lower, surrounding self
+	{
+		{ 1, -1, 0 },
+		{ -1, -1, 0 },
+		{ 0, -1, 1 },
+		{ 0, -1, -1 },
+	};
+	
+	for (auto Offset : LevelOffsets)
+	{
+		auto AdjustedPos = a_Entry.first + Offset;
+		auto Neighbour = a_Entry.second->GetRelNeighborChunkAdjustCoords(AdjustedPos.x, AdjustedPos.z);
+		auto MyPower = IsWirePowered(a_Entry.first, a_Entry.second);
 
-		BLOCKTYPE Type = 0;
-		if (!m_Chunk->UnboundedRelGetBlockType(a_RelBlockX + gCrossCoords[i].x, a_RelBlockY + gCrossCoords[i].y, a_RelBlockZ + gCrossCoords[i].z, Type))
+		if ((Neighbour == nullptr) || !Neighbour->IsValid())
 		{
-			continue;
+			return;
 		}
-		if (Type == E_BLOCK_REDSTONE_WIRE)
+
+		if ((Neighbour->GetBlock(AdjustedPos) == E_BLOCK_REDSTONE_WIRE) && (MyPower > 1) && (MyPower > IsWirePowered(AdjustedPos, Neighbour)))
 		{
-			SetBlockPowered(a_RelBlockX + gCrossCoords[i].x, a_RelBlockY + gCrossCoords[i].y, a_RelBlockZ + gCrossCoords[i].z, a_RelBlockX, a_RelBlockY, a_RelBlockZ, MyPower);
+			PowerBorderingWires(a_PotentialWireList, a_Entry, AdjustedPos, Neighbour, MyPower);
 		}
```

Comment:
This bock should be moved into a seperate method, so you don't repeat yourself for all three loops